### PR TITLE
[release/7.0] Unload MsQuic after checking for QUIC support to free resources (#75163)

### LINKED ISSUE #74629: HttpClient creates Quic threads even if HTTP/3 is not in use
1. Create simple console app with HTTP and run simple http1.1 request  on Windows 11 or Server 22. 
2. look at application threads.

image

Application crates 2x threads per core even if HTTP/3 and Quic are not in used. This was originally reported on legacy system with 54 cores. That will create 100+ threads ... for nothing. 

We should figure out how to make `Quic.IsSupported` cheap and defer thread creation until needed. 
This will impact legacy app that has no clue tor intention to use Quic or HTTP/3.

### PR DESCRIPTION
**Pull request by rzikm**
Manual backport of #75163
Fixes #74629

cc: @wfurt 

## Customer Impact

When using `HttpClient`, we also check whether the running platform supports QUIC (to enable HTTP/3). However, the way we are checking QUIC support causes many threads to be allocated in the native MsQuic library (2* number of logical cores). This causes unnecessary resource increase even if the process does not end up using HTTP/3 at all (HTTP/3 is opt-in). This would therefore cause regression in memory usage when upgrading to .NET 7 in such cases.

The change builds on fix in msquic -- therefore it includes also update of Docker images to version with the right msquic version (2.1.1).

Affected platforms include Windows 11, Windows Server 2022, many Linux platforms with msquic package installed. 

Note: The same problem exists also in 6.0 and we had 1 customer with ~50-core machine who noticed that and were surprised to see so many extra threads they didn't know about / they didn't need. At minimum it will help avoid confusion in such cases.

## Testing

Functional tests suite passes as part of the CI, resource consumption was checked manually.

## Risk

Low, the fix consists of gracefully unloading MsQuic library from the process after checking QUIC support. The library is reloaded only when actually needed.

### CODE REVIEW

**Review by karelz — APPROVED** [review]

**Comment by danmoseley** [comment]
Approved -- significant customer impact, new feature, and hitting existing customers that don't use it. Would certainly service for it.

But, more eyes would be good at this point. @stephentoub would it be possible for you to review before we merge here?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs:131]: I thought @jkotas had advised against actually unloading it. Did we decide it makes sense to anyway?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs:147]:
```C#
throw ThrowHelper.GetExceptionForMsQuicStatus(openStatus) ;
```
and then remove the next line?

**Comment by karelz** [comment]
Yeah, I have doubts that it may be cause by unloading msquic. Too high risk for not-so-large value. Let's hold off merging this PR. We will likely not take it into 7.0 at all.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicApi.cs:131]: I think we can keep the same benefit without unloading the library from the process, we just need to save the IntPtr handle to a static field. The threads should all get deallocated by `MsQuicClose`.

**Comment by danmoseley** [comment]
Closing per comments above.

**Comment by karelz** [comment]
FYI: The mysterious memory corruption was chased down to be unrelated problem in LTTNG triggered in our test environment - see #74795 for a test-only fix.

We have CR feedback addressed in #75441 -- calling `MsQuicClose` without unloading msquic DLL.

Given the churn in the area, I would like to see a little bit more bake time in main for the final changes. Once we have confidence it didn't break anything and is not triggering problems in LTTNG or elsewhere, I would be fine bringing it into 7.0. That will be already too late for GA, so we will leave it for servicing request if it confuses customers enough and if/when we get complains from customers.
@stephentoub @jkotas would you suggest something else?

**Comment by stephentoub** [comment]
I would like to see this fixed for .NET 7; it is a regression from .NET 6 (I just tried with .NET 6 on Win11, and without setting the experimental HTTP/3 flag, msquic is never loaded and the extra threads aren't created) that pretty much every app is going to end up paying even if HTTP/3 is never used.  I think it's fine to hold off for a few days or however long to ensure this actually addresses the problem without resulting in new instabilities, and it can be fixed post-RC2 for 7.0.0 or in servicing for 7.0.1.

**Comment by karelz** [comment]
Agreed in offline discussion. @rzikm feel free to reopen the PR and merge your PR-feedback. We will mark it NO-MERGE and decide how long we want to confirm it didn't cause any regression in main.

**Comment by jkotas** [comment]
> We will mark it NO-MERGE and decide how long we want to confirm it didn't cause any regression in main.

My preference would be to push the Quic fixes to release/7.0 sooner rather than later (e.g. after one day of bake time). Main is not getting a lot of scrutiny outside our CI system currently, so you are unlikely to get a lot more extra signal by waiting longer.

**Comment by rzikm** [comment]
I was unable to reopen this PR, so I created new at https://github.com/dotnet/runtime/pull/75521

**Pull request closed without merging**
